lib/gollum-lib/filter/yaml.rb

```diff
@@ -0,0 +1,26 @@
+require 'yaml'
+# Extract YAML frontmatter from data and build metadata table. 
+
+class Gollum::Filter::YAML < Gollum::Filter
+  
+  # Regexp thanks to jekyll
+  YAML_FRONT_MATTER_REGEXP = %r!\A(---\s*\n.*?\n?)^((---|\.\.\.)\s*$\n?)!m
+
+  def extract(data)
+    data.gsub!(YAML_FRONT_MATTER_REGEXP) do
+      @markup.metadata ||= {}
+      begin
+        frontmatter = ::YAML.load(Regexp.last_match[1])
+        frontmatter = frontmatter.respond_to?(:to_h) ? frontmatter.to_h : Hash.new(frontmatter)
```

Comment:
the `Hash.new` case is to deal with cases like this:
```ruby
2.4.1 :004 > YAML.load("---\ntrue\n---")
 => true  
```
This surprisingly doesn't throw a syntax error: as long as there is no `:` in the YAML block, it is simply parsed as a YAML object and returns a string, boolean, etc.

Calling `Hash.new(obj)` for any object that doesn't respond to `to_h` just results in an empty hash, so this is a way to make sure the metadata isn't accidentally set to an object of a different kind, without resorting to `is_a?(Hash)`. But I also don't think it's very transparent, so very open to other suggestions.

I also just realized there is this edge case, because `Array`s actually have a working `to_h` method :
```ruby
.4.1 :012 > YAML.load("---\n[[foo, bar],[waa,baa]]\n---").to_h
 => {"foo"=>"bar", "waa"=>"baa"} 
```